microsoft/onefuzz
# Abstract out node disposal

## Summary of the Pull Request

Introduces a new way for nodes to be reaped from a scale set. This allows azure auto scale to scale in nodes when appropriate.

## PR Checklist
* [x] Closes #1685

## Info on Pull Request

This PR includes:
* Abstracting out how nodes are disposed (azure auto-scale scale-in or node delete/reimage)
* The capability for onefuzz to correctly handle when azure scales the nodes in
* Removal of the `resize` state that is no longer necessary
* Tuning auto scale rules according to azure guidance

## Validation Steps Performed

1. Create a new scale set with some nodes
2. Do not submit any jobs
3. Watch auto scale delete the nodes
4. Verify the size of the scale set is accurately represented in azure table
5. Submit a job
6. Watch the scale set grow and items from the pool queue being consumed
7. Repeat step 4
8. Verify the protection policy on the busy nodes is set to "Scale in actions" in the "Instances" tab of the scale set
9. Once the job is complete, watch the nodes scale back in
